src/Features/Core/Portable/ConvertTupleToStruct/AbstractConvertTupleToStructCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,865 @@
+
+            return ImmutableArray.Create(new DocumentToUpdate(
+                document, ImmutableArray.Create(containingMember)));
+        }
+
+        private static async Task GenerateStructIntoContainingNamespaceAsync(
+            Document document, SyntaxNode tupleExprOrTypeNode, INamedTypeSymbol namedTypeSymbol, 
+            Dictionary<Document, SyntaxEditor> documentToEditorMap, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // If we don't already have an editor for the containing document, then make one.
+            if (!documentToEditorMap.TryGetValue(document, out var editor))
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                editor = new SyntaxEditor(root, generator);
+
+                documentToEditorMap.Add(document, editor);
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var container = tupleExprOrTypeNode.GetAncestor<TNamespaceDeclarationSyntax>() ?? root;
+
+            // Then, actually insert the new class in the appropriate container.
+            editor.ReplaceNode(container, (currentContainer, _) =>
+            {
+                var codeGenService = document.GetLanguageService<ICodeGenerationService>();
+                var options = new CodeGenerationOptions(
+                    generateMembers: true,
+                    sortMembers: false,
+                    autoInsertionLocation: false);
+
+                return codeGenService.AddNamedType(
+                    currentContainer, namedTypeSymbol, options, cancellationToken);
+            });
+        }
+
+        private async Task<Solution> ApplyChangesAsync(
+            Document startingDocument, Dictionary<Document, SyntaxEditor> documentToEditorMap, CancellationToken cancellationToken)
+        {
+            var currentSolution = startingDocument.Project.Solution;
+
+            foreach (var (currentDoc, editor) in documentToEditorMap)
+            {
+                var docId = currentDoc.Id;
+                var newRoot = editor.GetChangedRoot();
+                var updatedDocument = currentSolution.WithDocumentSyntaxRoot(docId, newRoot, PreservationMode.PreserveIdentity)
+                                                     .GetDocument(docId);
+
+                if (currentDoc == startingDocument)
+                {
+                    // If this is the starting document, format using the equals+getHashCode service
+                    // so that our generated methods follow any special formatting rules specific to
+                    // them.
+                    var equalsAndGetHashCodeService = startingDocument.GetLanguageService<IGenerateEqualsAndGetHashCodeService>();
+                    updatedDocument = await equalsAndGetHashCodeService.FormatDocumentAsync(
```

Comment:
>FormatDocumentAsync [](start = 72, length = 19)

Sorry if you've already answered this question before, but why format the entire document? Should we try to only format the generated code?

Reply:
No worries.  This won't actually format the entire document.  The docs for FormatDocument explain this:

```c#
    internal interface IGenerateEqualsAndGetHashCodeService : ILanguageService
    {
        /// <summary>
        /// Formats only the members in the provided document that were generated by this interface.
        /// </summary>
        Task<Document> FormatDocumentAsync(Document document, CancellationToken cancellationToken);
```